server/index-node/src/resolver.rs

```diff
@@ -149,24 +165,135 @@ where
+        let unvalidated_subgraph_manifest = {
+            let future = UnvalidatedSubgraphManifest::<graph_chain_ethereum::Chain>::resolve(
+                deployment_hash,
+                self.link_resolver.clone(),
+                &self.logger,
+            );
+            futures03::executor::block_on(future)
+                .map_err(|_error| QueryExecutionError::SubgraphManifestResolveError)?
+        };
+
+        // Feature management is not available for subgraphs with specVersion below 0.0.4
+        if *unvalidated_subgraph_manifest.spec_version() < SPEC_VERSION_0_0_4 {
+            return Err(QueryExecutionError::InvalidSubgraphManifest);
+        }
+
+        // Then validate the subgraph we've just obtained.
+        //
+        // Note that feature valiadation errors will be inside the error variant vector (because
+        // `validate` also validates subgraph features), so we must filter them out to build our
+        // response.
+        let subgraph_validation: Either<_, _> =
+            match unvalidated_subgraph_manifest.validate(self.subgraph_store.clone()) {
+                Ok((subgraph_manifest, _warnings)) => Either::Left(subgraph_manifest),
+                Err(validation_errors) => {
+                    if validation_errors.iter().all(|error| {
+                        matches!(
+                            error,
+                            SubgraphManifestValidationError::FeatureValidationError(_)
+                        )
+                    }) {
+                        Either::Right(validation_errors)
+                    } else {
+                        // If other errors are present, we must return early with an error.
+                        //
+                        // It might be useful to return a more thoughtful error, but that is not the
+                        // purpose of this endpoint.
+                        return Err(QueryExecutionError::InvalidSubgraphManifest);
+                    }
+                }
+            };
+
+        // At this point, we have either:
+        // 1. A valid subgraph manifest with no errors.
+        // 2. No subgraph manifest and a set of feature validation errors.
+        //
+        // For this step we must collect whichever results we have into GraphQL `Value` types.
+        let (features, errors) = match subgraph_validation {
+            Either::Left(subgraph_manifest) => {
+                let features = q::Value::List(
+                    detect_features(&subgraph_manifest)
+                        .map_err(|_| QueryExecutionError::InvalidSubgraphManifest)?
+                        .iter()
+                        .map(ToString::to_string)
+                        .map(q::Value::String)
+                        .collect(),
+                );
+                let errors = q::Value::List(vec![]);
+                (features, errors)
+            }
+            Either::Right(errors) => {
+                assert!(!errors.is_empty());
```

Comment:
What happens when confidence check fails? The GraphQL request receives what?

Reply:
The `IndexNodeServer` component panics, but since it is managed by a superior process, the error message gets both logged as an `ERROR` and returned in the response, like this:

```json
{
  "errors": [
    {
      "message": "panic processing query: Something went wrong [bla bla bla]..."
    }
  ]
}
```

Reply:
With the refactor, we don't need this confidence check anymore. Thanks @otaviopace!